recipes/mic.lua

```diff
@@ -0,0 +1,168 @@
+        * (c) 2021, bzgec
+
+
+# Microphone state widget/watcher
+
+This widget can be used to display the current microphone status.
+
+## Requirements
+
+- `amixer` - this command is used to get and toggle microphone state
+
+## Usage
+
+- Download [mic.lua](https://awesomewm.org/recipes/mic.lua) file and put it into awesome's
+  folder (like `~/.config/awesome/widgets/mic.lua`)
+
+- Add widget to `theme.lua`:
+
+```lua
+local widgets = {
+    mic = require("widgets/mic"),
+}
+theme.mic = widgets.mic({
+    timeout = 10,
+    settings = function(self)
+        if self.state == "muted" then
+            self.widget:set_image(theme.widget_micMuted)
+        else
+            self.widget:set_image(theme.widget_micUnmuted)
+        end
+    end
+})
+local widget_mic = wibox.widget { theme.mic.widget, layout = wibox.layout.align.horizontal }
+```
+
+- Create a shortcut to toggle microphone state (add to `rc.lua`):
+
+```lua
+-- Toggle microphone state
+awful.key({ modkey, "Shift" }, "m",
+          function ()
+              beautiful.mic:toggle()
+          end,
+          {description = "Toggle microphone (amixer)", group = "Hotkeys"}
+),
+```
+
+- You can also add a command to mute the microphone state on boot. Add this to your `rc.lua`:
+
+```lua
+-- Mute microphone on boot
+beautiful.mic:mute()
+```
+
+--]]
+
+
+local awful   = require("awful")
+local naughty = require("naughty")
+local gears   = require("gears")
+local wibox   = require("wibox")
+
+local function factory(args)
+    local args = args or {}
+
+    local mic = {
+        widget   = args.widget or wibox.widget.imagebox(),
+        settings = args.settings or function(self) end,
+        timeout  = args.timeout or 10,
+        timer    = gears.timer,
+        state    = "",
+    }
+
+    function mic:mute()
+        awful.spawn.easy_async_with_shell("amixer set Capture nocap",
+            function()
+                self:update()
+            end
+        )
+    end
+
+    function mic:unmute()
+        awful.spawn.easy_async_with_shell("amixer set Capture cap",
+            function()
+                self:update()
+            end
+        )
+    end
+
+    function mic:toggle()
+        awful.spawn.easy_async_with_shell("amixer set Capture toggle",
+            function()
+                self:update()
+            end
+        )
+    end
+
+    function mic:pressed(button)
+        if button == 1 then
+            self:toggle()
+        end
+    end
+
+    function mic:update()
+        -- Check that timer has started
+        if self.timer.started then
+            self.timer:emit_signal("timeout")
+        end
+    end
+
+    -- Read `amixer get Capture` command and try to `grep` all "[on]" lines.
+    --   - If there are lines with "[on]" then assume microphone is "unmuted".
+    --   - If there are NO lines with "[on]" then assume microphone is "muted".
+    mic, mic.timer = awful.widget.watch(
+        "bash -c \"amixer get Capture | grep '\\[on\\]'\"",
```

Comment:
Seems like a great place to use `{ "bash", "-c", "amixer get Capture | grep '\[on\]'" }` instead. Assuming this works, I think it is a lot more readable.

Also: In theory one could ditch the shell and instead do the grep in lua, but I can see why one wouldn't want to do that. :-)

Well... actually, you are just using grep to check whether some string is present. I *think* that `string.find(stdout, "[[]on]")` does that as well, but I can see that this is a little hard to read: `[[]` checks if any of the characters `[` are present. Similarly, `[ab]` would look for either `a` *or* `b`. I do not know how else to express this.

Reply:
> Seems like a great place to use `{ "bash", "-c", "amixer get Capture | grep '\[on\]'" }` instead. Assuming this works, I think it is a lot more readable.

You are right :)

> Also: In theory one could ditch the shell and instead do the grep in lua, but I can see why one wouldn't want to do that. :-)

True, but I like to check it this way so it is easily tested directly in terminal but more importantly could easily be ported to some other software.